core/controllers/reader_test.py

```diff
@@ -205,6 +257,106 @@ def test_huge_answers(self):
             'a' * 1000500, 'Sorry, nope, we didn\'t get it', '')
 
 
+class ReaderClassifyTests(test_utils.GenericTestBase):
+    """Test reader.classify using the sample explorations.
+
+    Since the end to end tests cover correct classification,
+    ReaderClassifyTests is only checking which of the hard/soft/classifier
+    rules is classified on input.
+    """
+
+    def setUp(self):
+        super(ReaderClassifyTests, self).setUp()
+        self._init_classify_inputs('16')
+
+    def _init_classify_inputs(self, exploration_id):
+        test_exp_filepath = os.path.join(
+            feconf.TESTS_DATA_DIR, 'string_classifier_test.yaml')
+        yaml_content = utils.get_file_contents(test_exp_filepath)
+        assets_list = []
+        exp_services.save_new_exploration_from_yaml_and_assets(
+            feconf.SYSTEM_COMMITTER_ID, yaml_content,
+            'Testing String Classifier', 'Test',
+            exploration_id, assets_list)
+
+        self.EXP_ID = exploration_id
+        self.EXP_STATE = (
+            exp_services.get_exploration_by_id(exploration_id).states['Home'])
+
+    def _get_classiying_rule_type(self, answer):
+        response = reader.classify(
+            self.EXP_ID, self.EXP_STATE, answer, {'answer': answer})
+        return response['matched_rule_type']
+
+    def test_hard_rule_classification(self):
```

Comment:
Okay, in general I am fine with these tests. However, they still feel more like integration tests to me. Sorry for mentioning so many suggestions with this test, but I think it's really important we thoroughly test one of the most important functions used by learners. Beyond that, this testing is going to be reflected in the frontend as Xinyu migrates this code over to the frontend, so it's really important to get it right now.

I think you should use your data exploration for the reader integration tests, like you are. For these specific reader.classify tests, I think you should be creating an in-memory exploration, populating it with just a few elements, and making sure reader.classify behaves *exactly* as it's supposed to. We should be able to read through this test suite and understand both how reader.classify works and what the exploration looks like which drives that behavior. Don't worry about the string classifier running correctly; you can use self.swap on the string classifier functions to verify they're called when they're supposed to be. There are references throughout exp_services_test and elsewhere demonstrating creating new explorations for testing purposes (test_utils.create_new_default_exploration and test_utils.create_new_valid_exploration will be helpful). If you have any questions or would like to discuss this more, I'll be happy to.